### PR TITLE
CA-280423: Open/close xapi-clusterd port when starting/stopping daemon

### PR DESCRIPTION
**Pull request by thomasmck**
Signed-off-by: Thomas Mckelvey <thomas.mckelvey@citrix.com>

### CODE REVIEW

File: ocaml/xapi/xapi_clustering.ml [hunk 1]
```diff
@@ -165,8 +165,10 @@ module Daemon = struct
     | None -> ignore (Helpers.call_script script params)
 
   let service = "xapi-clusterd"
+  let xapi_clusterd_port = "8896"
   let enable ~__context =
     debug "Enabling and starting the clustering daemon";
+    maybe_call_script ~__context "/etc/xapi.d/plugins/firewall-port" ["open"; xapi_clusterd_port];
```
Comment (robhoes) [inline ocaml/xapi/xapi_clustering.ml]: This path, and the other strings in here, should be defined in xapi_globs.ml, such that they can be set from the config file.

Comment (thomasmck) [reply ocaml/xapi/xapi_clustering.ml]: @robhoes Just to make sure I understand properly we should move the path, the port number and the "open"/"close" strings to xapi_globs.ml?

And then just call them here with `Xapi_globs.xapi_clusterd_port`?

Comment (robhoes) [reply ocaml/xapi/xapi_clustering.ml]: I think that you can leave "open"/"close", and just do the path and port. Grep for e.g. `nbd_firewall_config_script` in xapi_globs.ml to see how to refine them and register them with the config file logic.

Comment (thomasmck) [reply ocaml/xapi/xapi_clustering.ml]: Should I actually update xapi_conf to include the port? looks like a look of ports are included but commented out. Thats what I've done atm.

Comment (robhoes) [reply ocaml/xapi/xapi_clustering.ml]: That's handy yes. How about adding the other one as well?

Comment (robhoes) [reply ocaml/xapi/xapi_clustering.ml]: Btw, they are commented out, with the default value filled in, as examples in the config file. That is common practice for such files. So you've done the right thing.

Comment (thomasmck) [reply ocaml/xapi/xapi_clustering.ml]: I didn't see any of the other paths included in there but I can include the one I've added anyways. Seeing as it is commented out there is no harm.

File: ocaml/xapi/xapi_clustering.ml [hunk 2]
```diff
@@ -167,6 +167,7 @@ module Daemon = struct
   let service = "xapi-clusterd"
   let enable ~__context =
     debug "Enabling and starting the clustering daemon";
+    maybe_call_script ~__context !Xapi_globs.firewall_port_config_script ["open"; (string_of_int !Xapi_globs.xapi_clusterd_port)];
```
[the remark this answers is not in the document]
Comment (thomasmck) [reply ocaml/xapi/xapi_clustering.ml]: Sounds reasonable, I will amend it.

**Comment by robhoes** [comment]
Looks like a rebase went wrong on this branch: the PR now contains a bunch of unrelated commits.

**Comment by thomasmck** [comment]
Removed the other commits that got pulled in and made the other suggested changes.

File: ocaml/xapi/xapi_clustering.ml [hunk 3]
```diff
@@ -165,8 +165,10 @@ module Daemon = struct
     | None -> ignore (Helpers.call_script script params)
 
   let service = "xapi-clusterd"
+  let port = (string_of_int !Xapi_globs.xapi_clusterd_port)
```
Comment (lindig) [inline ocaml/xapi/xapi_clustering.ml]: Careful: the port is defined by xapi startup and any changes to !Xapi_globs.xapi_clusterd_port over the lifetime of xapi will not be taken into account. You might want to put this inside `enable()` and `disable()` to read it again at the time of usage.

Comment (robhoes) [reply ocaml/xapi/xapi_clustering.ml]: Indeed, `port` may evaluate before the config file read takes place.

File: scripts/xapi.conf [hunk 4]
```diff
@@ -378,6 +378,12 @@ sm-plugins=ext nfs iscsi lvmoiscsi dummy file hba rawhba udev iso lvm lvmohba lv
 # the vbd ring buffer
 # default-vbd3-polling-idle-threshold = 50
 
+# The communication port used by xapi-clusterd
+# xapi_clusterd_port = 8896
+
+# Path for the firewall-port script
+# firewall_port_config_script = /etc/xapi.d/plugins/firewall-port
```
Comment (robhoes) [inline scripts/xapi.conf]: `firewall-port-config`

**Comment by thomasmck** [comment]
Ok I think all comments have now been addressed. Let me know if there is anything I've missed.

**Comment by thomasmck** [comment]
I agree. Github has the ability to do that as part of the merge right? I do I need to do it myself and re-push?

**Comment by thomasmck** [comment]
@lindig @robhoes Do I need to squash manually myself or can it be done through github when merging?

**Pull request merged**